pymodbus/client/mixin.py

```diff
@@ -551,14 +551,21 @@ def convert_to_registers(
         """Convert int/float/str to registers (16/32/64 bit).
 
         :param value: value to be converted:
-        :param data_type: data type to convert to
+        :param data_type: data type to be encoded as registers
         :returns: List of registers, can be used directly in e.g. write_registers()
+        :raises TypeError: when there is a mismatch between data_type and value
         """
-        if data_type == cls.DATATYPE.STRING:
-            byte_list = value.encode()  # type: ignore[union-attr]
+        if "STRING" in data_type.name:
+            if not isinstance(value, str):
+                raise TypeError(f"Value should be string but is {type(value)}.")
+            byte_list = value.encode()
             if len(byte_list) % 2:
                 byte_list += b"\x00"
         else:
+            if "INT" in data_type.name and not isinstance(value, int):
```

Comment:
I have nothing against testing, but in that case you missed out on float. And I question if this test is worth while, because pack/unpack catches this.